lib/grandpa/grandpa.go

```diff
@@ -678,15 +678,61 @@ func (s *Service) deleteVote(key ed25519.PublicKeyBytes, stage Subround) {
 	}
 }
 
-// determinePreVote determines what block is our pre-voted block for the current round
-func (s *Service) determinePreVote() (*Vote, error) {
-	var vote *Vote
+func (s *Service) determineBestHeaderToPrevote(finalizedHeader, bestBlockHeader *types.Header) (
+	headerToPrevote *types.Header, err error) {
+	genesisHash := s.blockState.GenesisHash()
+	bestBlockHash := bestBlockHeader.Hash()
+
+	isGenesisHash := genesisHash.Equal(bestBlockHash)
+	if isGenesisHash || finalizedHeader.Hash().Equal(bestBlockHash) {
+		return bestBlockHeader, nil
+	}
 
+	isDescendant, err := s.blockState.IsDescendantOf(finalizedHeader.Hash(), bestBlockHash)
+	if err != nil {
+		return nil, fmt.Errorf("determining ancestry: %w", err)
+	}
+
+	if !isDescendant {
+		return nil, fmt.Errorf("%w: %s is not ancestor of %s",
+			blocktree.ErrNoCommonAncestor, bestBlockHash.Short(), bestBlockHash.Short())
+	}
+
+	// implements `BeforeBestBlockBy` a custom voting rule that guarantees that our vote is always
+	// behind the best block by at least N blocks, unless the base number is < N blocks behind the
+	// best, in which case it votes for the base.
+	// (https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/voting_rule.rs#L92)
```

Comment:
Should the code below be its own function with this comment above the function signature?